### ISSUE TITLE
Disable formatting

### ISSUE THREAD

**Issue by Nargonath** [typescript-language-server/typescript-language-server#257]
I'm using `nvim-lspconfig` and I want to disable `tsserver` formatting feature but I can't seem to find a way to do it through `typescript-language-server`. I skimmed through the issues but I couldn't find another question regarding this. Is there a way to pass options through `typescript-language-server` for that?

The only thing that I found that works is changing `resolved_capabilities` within a custom `on_attach` function. Here is what I have working and other failed attempts:

```lua
local nvim_lsp = require('lspconfig')

local on_attach = function(client, bufnr)
  -- only thing that works
  client.resolved_capabilities.document_formatting = false
end

local tsCapabilities = vim.lsp.protocol.make_client_capabilities()
-- doesn't work
tsCapabilities.textDocument.formatting = false

nvim_lsp.tsserver.setup({
  on_attach = on_attach,
  capabilities = tsCapabilities,
  settings = {
    -- doesn't work either
    documentFormatting = false
  }
})
```

**Comment by rchl** [comment]
There is currently no option in the server for that.

I'm not sure adding such option makes sense because the server just provides the functionality. It doesn't trigger formatting on its own. So I think it would be responsibility of the client to not trigger formatting if the user doesn't want it triggered.

It seems that you found a way to disable it already so doesn't that solve your issue already?
If you are looking for a simpler way of disabling it then maybe ask on the Neovim's repo?

**Comment by Nargonath** [comment]
Thanks for your prompt answer @rchl. I asked here because other LS seem to offer such feature i.e [`efm-langserver`](https://github.com/mattn/efm-langserver#initializeparams).

It's definitely not a blocker, I was just looking for a way to configure the LS I use in the same way.

I'll look in another direction. Thank you for your time. 😃

**Comment by David-Else** [comment]
@Nargonath Did you find a better solution to disable the formatting? It is very important for people who want to use Prettier... which seems to be everyone :)

Seems sublimelsp has it working somehow https://github.com/sublimelsp/LSP-typescript/issues/118#issuecomment-1004776231

**Comment by rchl** [comment]
It's the LSP client for ST that has a generic feature to disable server capabilities. Depending on what editor you are using, there might be something similar there.

**Comment by David-Else** [comment]
I ended up with the following inside the 'on_attach function', which is fine :)
```
  if client.name == 'jsonls' or 'tsserver' then
    client.resolved_capabilities.document_formatting = false
    client.resolved_capabilities.document_range_formatting = false
  end
```
Then I can set them both to prettier and job done!

**Comment by Nargonath** [comment]
Here is what I'm doing @David-Else:

```lua
local on_attach = function(client, bufnr)
	if client.name == "tsserver" then
		client.resolved_capabilities.document_formatting = false
		client.resolved_capabilities.document_range_formatting = false
	end
end
```

Then I pass that function to `on_attach` settings to nvim_lsp.

**Comment by mraaroncruz** [comment]
This is even used as an example in the neovim lsp docs.

https://neovim.io/doc/user/lsp.html#vim.lsp.buf.format%28%29

```lua
-- Never request typescript-language-server for formatting
vim.lsp.buf.format {
  filter = function(client) return client.name ~= "tsserver" end
}
```

**Comment by betoharres** [comment]
> This is even used as an example in the neovim lsp docs.
> 
> https://neovim.io/doc/user/lsp.html#vim.lsp.buf.format%28%29
> 
> -- Never request typescript-language-server for formatting
> vim.lsp.buf.format {
>   filter = function(client) return client.name ~= "tsserver" end
> }

sharing the complete code that prevents such annoying case for newbies stumbling on this issue and having more context:
```
local buffer_autoformat = function(bufnr)
  local group = 'lsp_autoformat'
  vim.api.nvim_create_augroup(group, {clear = false})
  vim.api.nvim_clear_autocmds({group = group, buffer = bufnr})
  vim.api.nvim_create_autocmd('BufWritePre', {
    buffer = bufnr,
    group = group,
    desc = 'LSP format on save',
    callback = function()
      -- Never request typescript-language-server for formatting
      vim.lsp.buf.format {
        async = false,
        timeout_ms = 10000,
        -- ts_ls formatting sucks
        filter = function(client) return client.name ~= "ts_ls" end
      }
    end,
  })
end

vim.api.nvim_create_autocmd('LspAttach', {
  callback = function(event)
    local id = vim.tbl_get(event, 'data', 'client_id')
    local client = id and vim.lsp.get_client_by_id(id)
    if client == nil then
      return
    end
    -- make sure there is at least one client with formatting capabilities
    if client.supports_method('textDocument/formatting') then
      buffer_autoformat(event.buf)
    end
  end
})

```
